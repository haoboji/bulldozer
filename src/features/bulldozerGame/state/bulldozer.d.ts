import { Terrain } from "./constant";

// Location: first number indicate x axis, second number indicate y axis
export type Location = [number, number];

// Direction: first number indicate x axis, second number indicate y axis
export type Direction = [number, number];

// Rotation: refer to https://en.wikipedia.org/wiki/Rotation_matrix
export type Rotation = [[number, number], [number, number]];

export interface Bulldozer {
  location: Location;
  direction: Direction;
}

export type Map = Terrain[][];

export interface Activity {
  terrain: Terrain;
  location: Location;
}
