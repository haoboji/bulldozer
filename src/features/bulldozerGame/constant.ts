export const PLAIN_LAND = "o";
export const ROCKY_LAND = "r";
export const REMOVABLE_TREE = "t";
export const PRESERVED_TREE = "T";
