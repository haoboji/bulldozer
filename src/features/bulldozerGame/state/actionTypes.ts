import { Rotation, SiteMap } from "./bulldozer";

export const SET_SITE_MAP = "SET_SITE_MAP";
export const ADVANCE_BULLDOZER = "ADVANCE_BULLDOZER";
export const ROTATE_BULLDOZER = "ROTATE_BULLDOZER";
export const END_SIMULATION = "END_SIMULATION";

export interface SetSiteMapAction {
  type: typeof SET_SITE_MAP;
  map: SiteMap;
}

export interface AdvanceBulldozerAction {
  type: typeof ADVANCE_BULLDOZER;
}

export interface RotateBulldozerAction {
  type: typeof ROTATE_BULLDOZER;
  rotation: Rotation;
}

export interface EndSimulationAction {
  type: typeof END_SIMULATION;
}

export type BulldozerGameAction =
  | SetSiteMapAction
  | AdvanceBulldozerAction
  | RotateBulldozerAction
  | EndSimulationAction;
