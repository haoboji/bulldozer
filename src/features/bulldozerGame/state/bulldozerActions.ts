import {
  AdvanceBulldozerAction,
  ADVANCE_BULLDOZER,
  EndSimulationAction,
  END_SIMULATION,
  RotateBulldozerAction,
  ROTATE_BULLDOZER,
} from "./actionTypes";
import { Rotation } from "./bulldozer";

export const advanceBulldozer = (): AdvanceBulldozerAction => ({
  type: ADVANCE_BULLDOZER,
});

export const rotateBulldozer = (rotation: Rotation): RotateBulldozerAction => ({
  type: ROTATE_BULLDOZER,
  rotation,
});

export const endSimuation = (): EndSimulationAction => ({
  type: END_SIMULATION,
});
