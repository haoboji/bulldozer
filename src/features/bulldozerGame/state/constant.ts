import { Direction, Rotation } from "./bulldozer";

export const DIRECTION_NORTH: Direction = [0, 1];
export const DIRECTION_SOUTH: Direction = [0, -1];
export const DIRECTION_EAST: Direction = [1, 0];
export const DIRECTION_WEST: Direction = [-1, 0];

// Rotation: refer to https://en.wikipedia.org/wiki/Rotation_matrix#Common_rotations
export const ROTATION_LEFT: Rotation = [
  [0, 1],
  [-1, 0],
];

export const ROTATION_RIGHT: Rotation = [
  [0, -1],
  [1, 0],
];

export enum Terrain {
  PlainLand = "o",
  RockyLand = "r",
  RemovableTree = "t",
  ProtectedTree = "T",
  ClearedLand = "c",
}

export enum GameStatus {
  Starting,
  Started,
  Error,
  Ended,
}

export enum Command {
  Advance,
  Left,
  Right,
  Quit,
}
