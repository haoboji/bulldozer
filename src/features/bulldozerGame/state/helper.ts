import { Location } from "./bulldozer";
import {
  PLAIN_LAND,
  PRESERVED_TREE,
  REMOVABLE_TREE,
  ROCKY_LAND,
} from "./constant";
import { SiteMap, Terrain } from "./site";

export const isLocationValid = (
  location: Location,
  map: SiteMap | null
): boolean => {
  const [x, y] = location;
  // Validate boundaries
  if (map?.[-y]?.[x] === undefined) {
    return false;
  }
  // Validate protected tree
  if (map[-y][x] === PRESERVED_TREE) {
    return false;
  }
  return true;
};

export const updateMapTile = (
  map: SiteMap,
  terrain: Terrain,
  rowIndex: number,
  colIndex: number
): SiteMap => {
  const row = map[rowIndex];
  const newRow = [
    ...row.slice(0, colIndex),
    terrain,
    ...row.slice(colIndex + 1),
  ];
  return [...map.slice(0, rowIndex), newRow, ...map.slice(rowIndex + 1)];
};

export const parseFile = (file: File): Promise<string> =>
  new Promise((resolve, reject) => {
    const fileReader = new FileReader();
    fileReader.onloadend = () => {
      resolve(fileReader.result as string);
    };
    fileReader.onerror = () => {
      reject(Error("Invalid file"));
    };
    fileReader.readAsText(file);
  });

export const parseMap = (content: string): SiteMap => {
  const trimmed = content.trim();
  if (trimmed.length === 0) {
    throw Error("Empty map");
  }
  const rawMap = trimmed.split("\n").map((row) => row.split(""));
  const abnormalRow = rawMap.find((r) => r.length !== rawMap[0].length);
  if (abnormalRow) {
    throw Error("Invalid map shape");
  }
  const invalidTile = rawMap.find(
    (row) =>
      !!row.find(
        (t) =>
          t !== PRESERVED_TREE &&
          t !== REMOVABLE_TREE &&
          t !== ROCKY_LAND &&
          t !== PLAIN_LAND
      )
  );
  if (invalidTile) {
    throw Error("Invalid map tile");
  }
  return rawMap as SiteMap;
};
