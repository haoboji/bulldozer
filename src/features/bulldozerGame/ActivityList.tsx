import { makeStyles, Paper } from "@material-ui/core";
import clsx from "clsx";
import React from "react";
import { FixedSizeList, ListChildComponentProps } from "react-window";
import config from "../../app/config";
import styles from "./ActivityList.styles";
import { PRESERVED_TREE } from "./state/constant";
import { Activity } from "./state/report";

export interface ActivityListProps {
  activities: Activity[];
}

const makeItems = (activities: ActivityListProps["activities"]) => {
  const Row = (props: ListChildComponentProps) => {
    const { style, index } = props;
    const classes = makeStyles(styles)();
    if (index === 0) {
      return (
        <div style={style} className={classes.item}>
          <span className={classes.name}>
            <b>Activity</b>
          </span>
          <span className={classes.location}>
            <b>Location ([x,y])</b>
          </span>
          <span className={classes.fuel}>
            <b>Fuel Usage (fuel unit)</b>
          </span>
          <span className={classes.cost}>
            <b>Cost (credit)</b>
          </span>
        </div>
      );
    }
    const activity = activities[index - 1];
    const { terrain, location } = activity;
    const { activityName, fuelUsage, itemCost } = config;
    const fuel = terrain !== PRESERVED_TREE ? fuelUsage[terrain] : 0;
    const cost = itemCost.fuel * fuel;

    return (
      <div style={style} className={clsx(classes.item, classes.borderTop)}>
        <span className={classes.name}>{activityName[terrain]}</span>
        <span className={classes.location}>{`[${location.join(",")}]`}</span>
        <span className={classes.fuel}>{fuel > 0 && fuel}</span>
        <span className={classes.cost}>{cost > 0 && cost}</span>
      </div>
    );
  };
  return Row;
};

const ActivityList = (props: ActivityListProps): JSX.Element => {
  const { activities } = props;
  const classes = makeStyles(styles)();

  return (
    <Paper variant="outlined" className={classes.root}>
      <FixedSizeList
        itemSize={36}
        width={650}
        height={420}
        itemCount={activities.length + 1}
      >
        {makeItems(activities)}
      </FixedSizeList>
    </Paper>
  );
};

export default ActivityList;
