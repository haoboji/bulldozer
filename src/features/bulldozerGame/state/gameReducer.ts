import {
  ADVANCE_BULLDOZER,
  BulldozerGameAction,
  END_SIMULATION,
  ROTATE_BULLDOZER,
  SET_SITE_MAP,
} from "./actionTypes";
import { Bulldozer, Location } from "./bulldozer";
import {
  Command,
  DIRECTION_EAST,
  GameStatus,
  ROTATION_RIGHT,
} from "./constant";
import { SiteMap } from "./site";
import { add, multiply } from "mathjs";
import { isLocationValid } from "./helper";
import { Activity } from "./report";

export interface GameState {
  map: SiteMap | null;
  status: GameStatus;
  bulldozer: Bulldozer;
  commands: Command[];
  activities: Activity[];
}

const initialBulldozer: Bulldozer = {
  location: [-1, 0],
  direction: DIRECTION_EAST,
};

export const initialGameState: GameState = {
  map: null,
  bulldozer: initialBulldozer,
  status: GameStatus.Starting,
  commands: [],
  activities: [],
};

const game = (
  state = initialGameState,
  action: BulldozerGameAction
): GameState => {
  switch (action.type) {
    case SET_SITE_MAP: {
      return {
        ...state,
        map: action.map,
        bulldozer: {
          location: [-1, action.map.length - 1],
          direction: DIRECTION_EAST,
        },
      };
    }
    case ADVANCE_BULLDOZER: {
      const { activities, bulldozer, map, commands } = state;
      const { location, direction } = bulldozer;
      const newLocation = add(location, direction) as Location;
      const [x, y] = newLocation;
      const terrain = map?.[map.length - y - 1]?.[x];
      const newActivities = terrain
        ? [
            ...activities,
            {
              terrain,
              location: newLocation,
            },
          ]
        : activities;
      const status = isLocationValid(newLocation, state.map)
        ? state.status
        : GameStatus.Error;
      return {
        ...state,
        status,
        activities: newActivities,
        commands: [...commands, Command.Advance],
        bulldozer: { ...bulldozer, location: newLocation },
      };
    }
    case ROTATE_BULLDOZER: {
      const { commands, bulldozer } = state;
      const { direction } = bulldozer;
      const { rotation } = action;
      const newDirection = multiply(direction, rotation);
      const command =
        rotation === ROTATION_RIGHT ? Command.Right : Command.Left;
      return {
        ...state,
        commands: [...commands, command],
        bulldozer: { ...bulldozer, direction: newDirection },
      };
    }
    case END_SIMULATION: {
      const { commands } = state;
      return {
        ...state,
        commands: [...commands, Command.Quit],
        status: GameStatus.Ended,
      };
    }
    default: {
      return state;
    }
  }
};

export default game;
