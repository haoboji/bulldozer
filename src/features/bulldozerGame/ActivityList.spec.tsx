import { render } from "@testing-library/react";
import React from "react";
import ActivityList from "./ActivityList";
import { PLAIN_LAND, REMOVABLE_TREE } from "./state/constant";
import { Activity } from "./state/report";

test("ActivityList component", () => {
  const activities: Activity[] = [
    { terrain: PLAIN_LAND, location: [12, 34] },
    { terrain: REMOVABLE_TREE, location: [56, 78] },
  ];
  const c = render(<ActivityList activities={activities} />);
  expect(c.getByText(/Clearing plain land/i)).toBeInTheDocument();
  expect(c.getByText(/Clearing land containing a tree/i)).toBeInTheDocument();
  expect(c.getByText("[12,34]")).toBeInTheDocument();
  expect(c.getByText("[56,78]")).toBeInTheDocument();
  expect(c.container).toMatchSnapshot();
});
