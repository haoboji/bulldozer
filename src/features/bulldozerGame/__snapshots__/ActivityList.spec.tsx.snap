// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityList component 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-root-1 MuiPaper-outlined MuiPaper-rounded"
  >
    <div
      style="position: relative; height: 420px; width: 650px; overflow: auto; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 108px; width: 100%;"
      >
        <div
          class="makeStyles-item-9"
          style="position: absolute; left: 0px; top: 0px; height: 36px; width: 100%;"
        >
          <span
            class="makeStyles-name-11"
          >
            <b>
              Activity
            </b>
          </span>
          <span
            class="makeStyles-location-12"
          >
            <b>
              Location ([x,y])
            </b>
          </span>
          <span
            class="makeStyles-fuel-13"
          >
            <b>
              Fuel Usage (fuel unit)
            </b>
          </span>
          <span
            class="makeStyles-cost-14"
          >
            <b>
              Cost (credit)
            </b>
          </span>
        </div>
        <div
          class="makeStyles-item-16 makeStyles-borderTop-17"
          style="position: absolute; left: 0px; top: 36px; height: 36px; width: 100%;"
        >
          <span
            class="makeStyles-name-18"
          >
            Clearing plain land
          </span>
          <span
            class="makeStyles-location-19"
          >
            [12,34]
          </span>
          <span
            class="makeStyles-fuel-20"
          >
            1
          </span>
          <span
            class="makeStyles-cost-21"
          >
            1
          </span>
        </div>
        <div
          class="makeStyles-item-23 makeStyles-borderTop-24"
          style="position: absolute; left: 0px; top: 72px; height: 36px; width: 100%;"
        >
          <span
            class="makeStyles-name-25"
          >
            Clearing land containing a tree
          </span>
          <span
            class="makeStyles-location-26"
          >
            [56,78]
          </span>
          <span
            class="makeStyles-fuel-27"
          >
            2
          </span>
          <span
            class="makeStyles-cost-28"
          >
            2
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
